testutils/infra/device.py

```diff
@@ -302,47 +301,68 @@ def _ssh_prep_args_impl(tool):
     return (cmd, host, port)
 
 
-def _put(file, local_path=".", remote_path="."):
-    (scp, host, port) = _scp_prep_args()
+def _put(device, file, local_path=".", remote_path="."):
+    (scp, host, port) = _scp_prep_args(device)
 
-    fabric.api.local(
+    subprocess.check_call(
         "%s %s %s/%s %s@%s:%s"
-        % (scp, port, local_path, file, fabric.api.env.user, host, remote_path)
+        % (scp, port, local_path, file, device.user, host, remote_path),
+        shell=True,
     )
 
 
-def _run(cmd, **kw):
+def _run(conn, cmd, **kw):
     if kw.get("wait") is not None:
         wait = kw["wait"]
         del kw["wait"]
     else:
         wait = 60 * 60
 
-    output = ""
+    result = None
     start_time = time.time()
     sleeptime = 1
-    # Use shorter timeout to get a faster cycle. Not recommended though, since
-    # in a heavily loaded environment, QEMU might be quite slow to use the
-    # connection.
-    with fabric.api.settings(timeout=60, abort_exception=Exception):
-        while True:
-            try:
-                output = fabric.api.run(cmd, **kw)
-                break
-            except Exception:
-                if time.time() >= start_time + wait:
-                    raise Exception("Could not successfully run: %s" % cmd)
-                time.sleep(sleeptime)
-                # Back off exponentially to save SSH handshakes in QEMU, which
-                # are quite expensive.
-                sleeptime *= 2
-                continue
-            finally:
-                # Taken from disconnect_all() in Fabric.
-                from fabric.state import connections
-
-                if connections.get(fabric.api.env.host_string) is not None:
-                    connections[fabric.api.env.host_string].close()
-                    del connections[fabric.api.env.host_string]
-
-    return output
+    while True:
```

Comment:
I can't see that this loop handles timing out, which the old one did. Should you perhaps check absolute time at the very beginning of the loop, before waiting?

Reply:
Yep, I realized this morning when I got the first tests running in CI. I have fixed it locally and will push soon. Thanks for a good finding!

Reply:
Pushed now, it has an extra handling for this:

```
        except UnexpectedExit as e:
            # Many tests rely on the old behaviour of "keep trying until it passes", so
            # they run commands that may return non 0 for a while
            logger.info(
                "Got UnexpectedExit while executing command in host %s: %s", conn.host, e
            )
            continue
```

It is not consistent across tests and I would love to clean-it up (many tests use "default" timeout which is ONE HOUR!). But I think I shouldn't spend too much time on this now so I added some TODOs and will address it in the future.

Reply:
> many tests use "default" timeout which is ONE HOUR!

Yes, this comes from the times we used Digital Ocean. It was not that uncommon that certain tests could take up to 40 min to pass. The capacity we got in Digital Ocean was absolutely abysmal. Several times I measured a steal time (time that hypervisor "steals" from VM) of **85%**! I think the conditions in Google Cloud are much better so maybe we could reduce this to something like 20 min or something. You could perhaps pick the longest test according to the html summary and multiply it by three or something. That should give a pretty safe fallback timeout, even when conditions are bad.